### ISSUE TITLE
Document how to deal with `SocketPermission` on `install_jar`

### ISSUE THREAD

**Issue by Zastai** [tada/pljava#425]
The documentation says

> The first is a URL to the jar file to be loaded. The simplest case would be a file: URL giving a path on the PostgreSQL server host to the jar file, with permissions set so the backend can read it. Other forms of URL are also accepted, so install_jar could retrieve a jar from the web, for example.

However, trying to use an https URL fails:
```sql
SELECT sqlj.install_jar('https://user:pass@initech.com/some-path/some.jar', 'some-jar', true);
```
results in `ERROR: java.sql.SQLSyntaxErrorException: access denied ("java.net.SocketPermission" "initech.com:443" "connect,resolve")`.

(this is with openjdk11 on debian running under WSL).

It's not clear to me where I need to configure something to enable such access (JDK/system level, pljava VM args, ...), so it would be useful to have this mentioned in the documentation.

**Comment by jcflack** [comment]
Thanks for the report. The place would be here in `pljava.policy`, which gets installed in `SYSCONFDIR`:

https://github.com/tada/pljava/blob/V1_6_4/pljava-packaging/src/main/resources/pljava.policy#L75

and it would surely be good for the documentation passage you found to mention it also.

**Comment by Zastai** [comment]
Is there any tracing that confirms whether that policy file is being used?
When is it (re)read (i.e. is a postgres restart needed when it is modified, or will the next use of pljava pick it up)?

I tried adding
```
permission java.net.SocketPermission "initech.com:443", "connect,resolve";
```
to `pljava.policy` (did have to chmod it to 644; it was 444 by default), in the section you indicated, after the read-all-files permission, but even after restarting postgres, an install_jar reported the same error as above.

I tried using
```
permission java.net.SocketPermission "*", "connect,resolve";
```
instead, and putting the permission before the file permission, but that made no difference.

Adding it to the global grants did have an effect however(*), so I guess Java does not think the socket access comes from the PL/Java code.

(*) well I'm getting a 401 despite having `https://user:pass@host/`, so I guess accessing URLs that require authorization (which I would assume would not be an uncommon scenario for JAR files) is not something that will work with the current API (Java seems to require setting up an Authenticator for that, so there might need to be an overload of install_jar that takes a username and password for that to work).

**Comment by Zastai** [comment]
It also looks like names like `foo-bar` are not allowed for the jar (which is a shame, because that's the typical naming convention for the jar files themselves); I can use `foo_bar` instead, but that's perhaps also something that bears explicit mention in the documentation.

**Comment by Zastai** [comment]
The `URLPermission` without corresponding `SocketPermission` does seem to work (but still with the 401 for my case).

**Comment by jcflack** [comment]
It is good to know the `URLPermission` works (and, I trust, works when placed in the grant clause specific to o.p.pljava.codesource?).

It looks as though the authentication support just needs to be added in `addClassImages`, which should be straightforward. It is disappointing that Java natively _parses and recognizes_ a _userinfo_ part of a URL, but does not _use_ it without further effort.

**Comment by Zastai** [comment]
Argh, I'll have to retest this tomorrow. While I did add it in the codesource section (and commented out the `SocketPermission` entry there), I failed to spot the global-level `SocketPermission` I added.

**Comment by Zastai** [comment]
OK, so when there is only a `URLPermission`, at the codebase grant level, things do not work (with the same error about `SocketPermission` being needed with `connect,resolve`.
Same when adding the `URLPermission` at the top level.

So it looks like the error is accurate, and a `SocketPermission` is needed. And I assume that you are using an API that just reads from a URI, with the specific scheme (file/ftp/http/...) being irrelevant, so the actual socket access is not in your code, which means it makes sense that the permission needs to be given elsewhere.

**Comment by jcflack** [comment]
That is quite possibly my fault. I may have been too clever by half in [dropping permissions][dprm] to only what the `URLConnection` says it needs.

A `URLPermission` (added in Java 1.8) ought to do the trick, but when you ask an `HttpURLConnection` what permission it needs ([older method][gp]) it replies with a `SocketPermission`.

So if starting with a `URLPermission` and then making the connection with permissions dropped to just what the `HttpUrlConnection` says it needs, the connection may end up being attempted with permissions of ∅.

I probably need to give up on the permission-dropping, and use the non-restricting form of `doPrivileged` there.

Hmm ... how to get some `install_jar`-with-a-URL tests into the CI script ....

[dprm]: https://github.com/tada/pljava/blob/V1_6_4/pljava/src/main/java/org/postgresql/pljava/management/Commands.java#L411
[gp]: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/net/HttpURLConnection.html#getPermission()

**Comment by jcflack** [comment]
`install_jar`-with-http-URL tests, that is. The `URLConnection` for a `file:` URL of course responds with a `FilePermission` narrowed to the specific file, and that works fine.

As nothing happens in the `doPrivileged` block beyond getting an `InputStream` from the URL for further processing outside of it, there should be little opportunity for mischief using the non-restricting form there.

**Comment by jcflack** [comment]
`MyNefariousStreamHandler` could also return a `NefariousURLConnection` whose `getPermission` method says "I require `AllPermission`, thank you very much", so the exercise of narrowing permissions to what the `URLConnection` requires wouldn't help there.

So yes, `RuntimePermission "setFactory"` should not be doled out like candy.

**Comment by jcflack** [comment]
Would you have an opportunity to try a build from the [bug/REL1_6_STABLE/issue425][branch] branch and see if this works for you? I have added a test of `install_jar` over HTTP to the CI script and it works for me locally. Again, Appveyor is showing all failures, which appear to be some issue between Appveyor and Maven repositories rather than a PL/Java problem.

In the CI script, a `URLPermission` is granted to PL/Java's codebase using the trial policy feature. You could grant a similar permission in your normal policy, and make it narrower than `http:*` if you like.

[branch]: https://github.com/tada/pljava/commits/bug/REL1_5_STABLE/issue425

**Comment by Zastai** [comment]
Will do, although I currently don't know when that would be.

**Comment by jcflack** [comment]
Believed resolved in 1.6.5.
